File: mcp/server/typescript/examples.ts
/**
 * Example Extensions for the Clean MCP Server (TypeScript)
 * 
 * This file contains examples of how to extend the base server with additional
 * tools and resources. Copy these examples into your server.ts file to add
 * new functionality.
 * 
 * These are just examples - participants can use them as inspiration!
 */

import fs from 'fs/promises';
import path from 'path';
import os from 'os';

// =============================================================================
// Example 1: Calculator Tool
// =============================================================================

// Add this interface near the top of your server.ts file:
interface CalculatorArgs {
  expression: string;
}

// Add this to your tools list in ListToolsRequestSchema handler:
const calculatorTool = {
  name: "calculator",
  description: "Perform basic math calculations (supports +, -, *, /, parentheses)",
  inputSchema: {
    type: "object",
    properties: {
      expression: {
        type: "string",
        description: "Mathematical expression to evaluate (e.g., '2 + 3 * 4')"
      }
    },
    required: ["expression"]
  }
};

// Add this case to your CallToolRequestSchema handler:
/*
case "calculator":
  try {
    const { expression } = args as CalculatorArgs;
    
    // Simple math evaluation (be careful with eval in production!)
    // In production, use a proper math parser like math.js
    const result = Function(`"use strict"; return (${expression})`)();
    
    return {
      content: [
        {
          type: "text",
          text: `Calculator Result:\n${expression} = ${result}`
        }
      ]
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Error calculating '${(args as CalculatorArgs).expression}': ${error instanceof Error ? error.message : String(error)}`
        }
      ]
    };
  }
*/

// =============================================================================
// Example 2: File System Tools
// =============================================================================

interface ListFilesArgs {
  path?: string;
}

interface ReadFileArgs {
  filePath: string;
}

// Add these tools to your tools list:
const fileSystemTools = [
  {
    name: "list_files",
    description: "List files and directories in a specified path",
    inputSchema: {
      type: "object",
      properties: {
        path: {
          type: "string",
          description: "Directory path to list (default: current directory)",
          default: "."
        }
      }
    }
  },
  {
    name: "read_file",
    description: "Read the contents of a text file",
    inputSchema: {
      type: "object",
      properties: {
        filePath: {
          type: "string",
          description: "Path to the file to read"
        }
      },
      required: ["filePath"]
    }
  }
];

// Add these cases to your CallToolRequestSchema handler:
/*
case "list_files":
  try {
    const { path: dirPath = "." } = args as ListFilesArgs;
    const files = await fs.readdir(dirPath);
    
    const fileDetails = await Promise.all(
      files.map(async (file) => {
        const fullPath = path.join(dirPath, file);
        const stats = await fs.stat(fullPath);
        return {
          name: file,
          isDirectory: stats.isDirectory(),
          size: stats.size,
          modified: stats.mtime.toISOString()
        };
      })
    );
    
    const fileList = fileDetails
      .map(f => `${f.isDirectory ? '📁' : '📄'} ${f.name} ${f.isDirectory ? '' : `(${f.size} bytes)`}`)
      .join('\n');
    
    return {
      content: [
        {
          type: "text",
          text: `Files in '${dirPath}':\n${fileList}`
        }
      ]
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Error listing files: ${error instanceof Error ? error.message : String(error)}`
        }
      ]
    };
  }

case "read_file":
  try {
    const { filePath } = args as ReadFileArgs;
    const content = await fs.readFile(filePath, 'utf-8');
    
    return {
      content: [
        {
          type: "text",
          text: `Contents of '${filePath}':\n\n${content}`
        }
      ]
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Error reading file '${(args as ReadFileArgs).filePath}': ${error instanceof Error ? error.message : String(error)}`
        }
      ]
    };
  }
*/

// =============================================================================
// Example 3: HTTP API Integration
// =============================================================================

interface WeatherArgs {
  city: string;
  apiKey?: string;
}

// Add this tool to your tools list:
const weatherTool = {
  name: "get_weather",
  description: "Get current weather information for a city",
  inputSchema: {
    type: "object",
    properties: {
      city: {
        type: "string",
        description: "City name to get weather for"
      },
      apiKey: {
        type: "string",
        description: "OpenWeatherMap API key (optional if set in environment)"
      }
    },
    required: ["city"]
  }
};

// Add this case to your CallToolRequestSchema handler:
/*
case "get_weather":
  try {
    const { city, apiKey } = args as WeatherArgs;
    const key = apiKey || process.env.OPENWEATHER_API_KEY;
    
    if (!key) {
      return {
        content: [
          {
            type: "text",
            text: "Error: No API key provided. Please provide apiKey parameter or set OPENWEATHER_API_KEY environment variable."
          }
        ]
      };
    }
    
    const url = `https://api.openweathermap.org/data/2.5/weather?q=${encodeURIComponent(city)}&appid=${key}&units=metric`;
    const response = await fetch(url);
    
    if (!response.ok) {
      throw new Error(`Weather API request failed: ${response.status}`);
    }
    
    const data = await response.json();
    const weather = {
      city: data.name,
      country: data.sys.country,
      temperature: Math.round(data.main.temp),
      description: data.weather[0].description,
      humidity: data.main.humidity,
      windSpeed: data.wind.speed
    };
    
    return {
      content: [
        {
          type: "text",
          text: `Weather in ${weather.city}, ${weather.country}:
🌡️ Temperature: ${weather.temperature}°C
🌤️ Conditions: ${weather.description}
💧 Humidity: ${weather.humidity}%
💨 Wind Speed: ${weather.windSpeed} m/s`
        }
      ]
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Error getting weather: ${error instanceof Error ? error.message : String(error)}`
        }
      ]
    };
  }
*/

// =============================================================================
// Example 4: Text Processing Tool
// =============================================================================

interface TextAnalysisArgs {
  text: string;
}

// Add this tool to your tools list:
const textAnalysisTool = {
  name: "analyze_text",
  description: "Analyze text for various metrics (word count, readability, etc.)",
  inputSchema: {
    type: "object",
    properties: {
      text: {
        type: "string",
        description: "Text to analyze"
      }
    },
    required: ["text"]
  }
};

// Add this case to your CallToolRequestSchema handler:
/*
case "analyze_text":
  try {
    const { text } = args as TextAnalysisArgs;
    
    const analysis = {
      characters: text.length,
      charactersNoSpaces: text.replace(/\s/g, '').length,
      words: text.trim().split(/\s+/).filter(word => word.length > 0).length,
      sentences: text.split(/[.!?]+/).filter(s => s.trim().length > 0).length,
      paragraphs: text.split(/\n\s*\n/).filter(p => p.trim().length > 0).length,
      lines: text.split('\n').length
    };
    
    const avgWordsPerSentence = analysis.sentences > 0 ? (analysis.words / analysis.sentences).toFixed(1) : '0';
    const avgCharsPerWord = analysis.words > 0 ? (analysis.charactersNoSpaces / analysis.words).toFixed(1) : '0';
    
    return {
      content: [
        {
          type: "text",
          text: `📊 Text Analysis Results:
━━━━━━━━━━━━━━━━━━━━━━━━━━
📏 Characters: ${analysis.characters.toLocaleString()}
📏 Characters (no spaces): ${analysis.charactersNoSpaces.toLocaleString()}
📝 Words: ${analysis.words.toLocaleString()}
📋 Sentences: ${analysis.sentences.toLocaleString()}
📄 Paragraphs: ${analysis.paragraphs.toLocaleString()}
📰 Lines: ${analysis.lines

File: mcp/server/python/examples.py
"""
Example Extensions for the Clean MCP Server

This file contains examples of how to extend the base server with additional
tools and resources. Copy these examples into your server.py file to add
new functionality.

These are just examples - participants can use them as inspiration!
"""

import json
import os
import datetime
from typing import Any
import aiohttp
import asyncio

# Example 1: Calculator Tool
# Add this to your handle_list_tools() function:
"""
Tool(
    name="calculator",
    description="Perform basic math calculations",
    inputSchema={
        "type": "object",
        "properties": {
            "expression": {
                "type": "string",
                "description": "Mathematical expression to evaluate (e.g., '2 + 3 * 4')",
            }
        },
        "required": ["expression"],
    },
),
"""

# Add this case to your handle_call_tool() function:
"""
elif name == "calculator":
    expression = arguments.get("expression", "")
    try:
        # Simple evaluation - in production, use a proper math parser
        result = eval(expression)  # WARNING: Don't use eval() in production!
        return [
            TextContent(
                type="text",
                text=f"Result: {expression} = {result}",
            )
        ]
    except Exception as e:
        return [
            TextContent(
                type="text",
                text=f"Error calculating '{expression}': {str(e)}",
            )
        ]
"""


# Example 2: File System Tool
# Add this to handle_list_tools():
"""
Tool(
    name="list_files",
    description="List files in a directory",
    inputSchema={
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": "Directory path to list (default: current directory)",
                "default": "."
            }
        },
    },
),
"""

# Add this case to handle_call_tool():
"""
elif name == "list_files":
    path = arguments.get("path", ".")
    try:
        files = os.listdir(path)
        file_list = "\n".join([f"{'📁' if os.path.isdir(os.path.join(path, f)) else '📄'} {f}" for f in sorted(files)])
        return [
            TextContent(
                type="text",
                text=f"Files in '{path}':\n{file_list}",
            )
        ]
    except Exception as e:
        return [
            TextContent(
                type="text",
                text=f"Error listing files in '{path}': {str(e)}",
            )
        ]
"""


# Example 3: Weather API Tool (requires aiohttp)
# Add this to handle_list_tools():
"""
Tool(
    name="get_weather",
    description="Get weather information for a city",
    inputSchema={
        "type": "object",
        "properties": {
            "city": {
                "type": "string",
                "description": "City name to get weather for",
            }
        },
        "required": ["city"],
    },
),
"""

# Add this case to handle_call_tool():
"""
elif name == "get_weather":
    city = arguments.get("city", "")
    try:
        # Using OpenWeatherMap API (you'd need an API key)
        # This is just an example structure
        async with aiohttp.ClientSession() as session:
            # Replace YOUR_API_KEY with actual API key
            url = f"http://api.openweathermap.org/data/2.5/weather?q={city}&appid=YOUR_API_KEY&units=metric"
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    temp = data['main']['temp']
                    description = data['weather'][0]['description']
                    return [
                        TextContent(
                            type="text",
                            text=f"Weather in {city}: {temp}°C, {description}",
                        )
                    ]
                else:
                    return [
                        TextContent(
                            type="text",
                            text=f"Could not get weather for {city}",
                        )
                    ]
    except Exception as e:
        return [
            TextContent(
                type="text",
                text=f"Error getting weather: {str(e)}",
            )
        ]
"""


# Example 4: System Information Resource
# Add this to handle_list_resources():
"""
Resource(
    uri="system://info",
    name="System Information",
    description="Current system information",
    mimeType="application/json",
),
"""

# Add this case to handle_read_resource():
"""
elif uri == "system://info":
    system_info = {
        "timestamp": datetime.datetime.now().isoformat(),
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        "platform": os.name,
        "current_directory": os.getcwd(),
        "environment_variables": dict(os.environ),
    }
    return json.dumps(system_info, indent=2)
"""


# Example 5: Text Processing Tool
# Add this to handle_list_tools():
"""
Tool(
    name="text_analyzer",
    description="Analyze text (word count, character count, etc.)",
    inputSchema={
        "type": "object",
        "properties": {
            "text": {
                "type": "string",
                "description": "Text to analyze",
            }
        },
        "required": ["text"],
    },
),
"""

# Add this case to handle_call_tool():
"""
elif name == "text_analyzer":
    text = arguments.get("text", "")
    
    # Basic text analysis
    word_count = len(text.split())
    char_count = len(text)
    char_count_no_spaces = len(text.replace(" ", ""))
    sentence_count = len([s for s in text.split(".") if s.strip()])
    paragraph_count = len([p for p in text.split("\n\n") if p.strip()])
    
    analysis = f'''Text Analysis Results:
━━━━━━━━━━━━━━━━━━━━
📊 Characters: {char_count}
📊 Characters (no spaces): {char_count_no_spaces}
📊 Words: {word_count}
📊 Sentences: {sentence_count}
📊 Paragraphs: {paragraph_count}

📈 Average words per sentence: {word_count/max(sentence_count, 1):.1f}
📈 Average characters per word: {char_count_no_spaces/max(word_count, 1):.1f}
'''
    
    return [
        TextContent(
            type="text",
            text=analysis,
        )
    ]
"""


# Example 6: Configuration Resource
# Add this to handle_list_resources():
"""
Resource(
    uri="config://settings",
    name="Server Configuration",
    description="Current server configuration and settings",
    mimeType="application/json",
),
"""

# Add this case to handle_read_resource():
"""
elif uri == "config://settings":
    config = {
        "server_name": "clean-mcp-server",
        "version": "1.0.0",
        "features": {
            "tools_enabled": True,
            "resources_enabled": True,
            "logging_level": "INFO"
        },
        "supported_transports": ["stdio"],
        "max_request_size": "1MB",
        "created": "2024-01-01T00:00:00Z"
    }
    return json.dumps(config, indent=2)
"""


# Workshop Exercise Ideas:
"""
🎯 Workshop Challenges for Participants:

1. **Easy Level:**
   - Add a "reverse_text" tool that reverses input text
   - Create a "current_time" resource that shows the current date/time
   - Build a "random_number" tool that generates random numbers

2. **Medium Level:**
   - Implement a simple task/todo manager with add/list/remove tools
   - Create a text file reader/writer with proper error handling
   - Build a URL shortener or QR code generator tool

3. **Advanced Level:**
   - Integrate with a real API (weather, news, stock prices)
   - Create a mini database using JSON files for persistence
   - Build a web scraper tool with BeautifulSoup
   - Implement a chat bot that connects to an AI API

4. **Expert Level:**
   - Create a multi-tool server that connects to multiple APIs
   - Build a data pipeline that processes CSV/JSON files
   - Implement real-time data streaming from external sources
   - Create a server that can spawn and manage other MCP servers
"""
